protoc-gen-go/testdata/servicedescriptor_test.go

```diff
@@ -0,0 +1,44 @@
+package testdata
+
+import (
+	"testing"
+	"google.golang.org/grpc"
+)
+
+func TestServiceDescriptor(t *testing.T) {
+	desc := GetTestServiceDesc()
+	if desc.ServiceName != "testdata.Test" {
+		t.Error("Service name does not match")
+	}
+	if len(desc.Methods) != 1 {
+		t.Error("Expected 1 method")
+	}
+	if len(desc.Streams) != 3 {
+		t.Error("Expected 3 streams")
+	}
+	if expectedDescString != ListMethodsAndStreams(desc) {
```

Comment:
Marshaling the descriptor into a string and then testing is not testing the actual public API or contract.

One should likely check that the ServiceName, MethodName(s), and []interface{}{ StreamName, ClientStream, ServerStream }... are all correct directly (since the func values preclude reflect.DeepEqual from working). Oh, and checking the Handlers to be non-nil would also be a good idea.

As a trivial (and dumb) bad-data-passes-test example, what if the Descriptor returned with two StreamNames, "Downstream\n\ttestdata.Test/Upstream" and "Bidi"?

“That would never happen!” But this is what testing is here for. We want to test the contract, (i.e. that the proper number and value of Descriptors is returned) not that marshaling it together in this specific way produces the same string. (n.b.: this example is ridiculous because it is a trivial set of bad data that produces a passing test… there could be further non-trivial bad data that would still pass, but are harder to think of.)

Reply:
I agree. Furthermore, this test would be more appropriate if it was table-driven. Provide expected service descriptors and check to see if the one returned from Get*ServiceDesc are equivalent. This will avoid this test getting out of date as Get*ServiceDesc hides or changes internals.

Some other tests that could be useful here:
- Checking that an imported proto package returns valid ServiceDescriptor objects
- Checking that modifying a given ServiceDescriptor before registering it affects the related service's provided API. I.E., can you call the renamed methods if you get a service descriptor and modify the names/positions/add a method?

Reply:
"table-driven" is overrated.  In a table-driven test you get all of the pass tests and all of the fail tests in the same *_test.go files and line reference.

This is solely and only useful when a piece of code has such an unreasonably wide scope that it has to be tested by algorithmic production. i.e. metaprogramming. Sometimes this is unavoidable, but when a specific file is supposed to produce a singular canonical output, that output should be known ahead of time, and should be human generatable, or the results themselves become themselves subject to testing, lest we become like Windows' "the implementation defines the behavior".

Reply:
I agree table-driven tests are overrated. I also use table-drive tests a lot.

I'm not really testing anything here because @paranoiacblack asked me to provide a test case as an example of usage. This is an example of usage. I'd like to be able to generate a list of methods that a protobuf service provides as a string, to display to a command-line user.

If you are looking unit tests for regressions, for documentation, for corner cases, etc, I'm happy to spend the time to do that, _after_ the discussion https://github.com/golang/protobuf/issues/489 finishes and I can be sure I'm not wasting my time here.

Reply:
Let's move forward with some sort of unit testing, even if it isn't very good. I don't get the segue into table-drive tests and it doesn't matter to me. Write a good test that isn't table-driven or write a good test that is table-driven. It's not very interesting to bikeshed this little thing but if you write a unit test that repeats a bunch of logic, I'll probably make you rewrite it.